src/Checkout/Helpers/ReserveStock.php

```diff
@@ -0,0 +1,176 @@
+<?php
+/**
+ * Handle product stock reservation during checkout.
+ *
+ * @package Automattic/WooCommerce
+ */
+
+namespace Automattic\WooCommerce\Checkout\Helpers;
+
+defined( 'ABSPATH' ) || exit;
+
+/**
+ * Stock Reservation class.
+ */
+final class ReserveStock {
+	/**
+	 * Query for any existing holds on stock for this item.
+	 *
+	 * @param \WC_Product $product Product to get reserved stock for.
+	 * @param integer     $exclude_order_id Optional order to exclude from the results.
+	 *
+	 * @return integer Amount of stock already reserved.
+	 */
+	public function get_reserved_stock( \WC_Product $product, $exclude_order_id = 0 ) {
+		global $wpdb;
+
+		// phpcs:ignore WordPress.DB.PreparedSQL.InterpolatedNotPrepared, WordPress.DB.PreparedSQL.NotPrepared
+		return (int) $wpdb->get_var( $this->get_query_for_reserved_stock( $product->get_stock_managed_by_id(), $exclude_order_id ) );
+	}
+
+	/**
+	 * Put a temporary hold on stock for an order if enough is available.
+	 *
+	 * @throws ReserveStockException If stock cannot be reserved.
+	 *
+	 * @param \WC_Order $order Order object.
+	 * @param int       $minutes How long to reserve stock in minutes. Defaults to woocommerce_hold_stock_minutes.
+	 */
+	public function reserve_stock_for_order( \WC_Order $order, $minutes = 0 ) {
+		$minutes = $minutes ? $minutes : (int) get_option( 'woocommerce_hold_stock_minutes', 60 );
+
+		if ( ! $minutes ) {
+			return;
+		}
+
+		try {
+			$items = array_filter(
+				$order->get_items(),
+				function( $item ) {
+					return $item->is_type( 'line_item' ) && $item->get_product() instanceof \WC_Product && $item->get_quantity() > 0;
+				}
+			);
+			$rows  = array();
+
+			foreach ( $items as $item ) {
+				$product = $item->get_product();
+
+				if ( ! $product->is_in_stock() ) {
+					throw new ReserveStockException(
+						'product_out_of_stock',
+						sprintf(
+							/* translators: %s: product name */
+							__( '%s is out of stock and cannot be purchased.', 'woocommerce' ),
+							$product->get_name()
+						),
+						403
+					);
+				}
+
+				// If stock management is off, no need to reserve any stock here.
+				if ( ! $product->managing_stock() || $product->backorders_allowed() ) {
+					continue;
+				}
+
+				$managed_by_id          = $product->get_stock_managed_by_id();
+				$rows[ $managed_by_id ] = isset( $rows[ $managed_by_id ] ) ? $rows[ $managed_by_id ] + $item->get_quantity() : $item->get_quantity();
+			}
+
+			if ( ! empty( $rows ) ) {
+				foreach ( $rows as $product_id => $quantity ) {
+					$this->reserve_stock_for_product( $product_id, $quantity, $order, $minutes );
+				}
+			}
+		} catch ( ReserveStockException $e ) {
+			$this->release_stock_for_order( $order );
+			throw $e;
+		}
+	}
+
+	/**
+	 * Release a temporary hold on stock for an order.
+	 *
+	 * @param \WC_Order $order Order object.
+	 */
+	public function release_stock_for_order( \WC_Order $order ) {
+		global $wpdb;
+
+		$wpdb->delete(
+			$wpdb->wc_reserved_stock,
```

Comment:
note that there is a chance that db migrations have not been run yet and tables are unavailable.

Reply:
How should this is mitigated? In this PR?

Reply:
yup, I think we need to have this behavior inside a feature flag and enable it only when we know migrations are completed. otherwise, the checkout will stop working for merchants who have not run the data migration yet.

Reply:
@vedanshujain Wait, unless this has changed recently, the installer runs on install/update/activate:

https://github.com/woocommerce/woocommerce/blob/6a8d8b27d98fe3036f2240d14be2b7a5e4ecf10e/includes/class-wc-install.php#L667

This is separate from the upgrade system and migration which requires user intervention. The tables would only be missing on error e.g. user lacked permissions to create tables.

Reply:
You are right, I tested this out and as long version is increased, tables should get installed in first request. Resolving this conversation.